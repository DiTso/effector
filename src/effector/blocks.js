//@flow

import {step, readRef} from 'effector/stdlib'

//prettier-ignore
export const filterChanged = step.filter({
  fn: (data, {state}) => (
    data !== readRef(state)
    && data !== undefined
  ),
})

export const noop = step.compute({
  fn: n => n,
})

export const updateHandler = step.compute({
  fn: (upd, {handler}) => handler(upd),
})
