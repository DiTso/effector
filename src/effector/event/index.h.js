//@flow
import type {Subscription, Subscriber} from '../index.h'
import type {kind, Graph, Unit} from 'effector/stdlib'
import type {CompositeName} from '../compositeName'

export type Event<E> = /*::interface extends Unit*/ {
  /*::
  [[call]](payload: E): E,
  */
  /*::+*/ id: string,
  getType(): string,
  create(payload: E, type: string, args: any[]): E,
  watch(watcher: (payload: E) => any): Subscription,
  map<T>(fn: (_: E) => T): Event<T>,
  filter<T>(fn: (_: E) => T | void): Event<T>,
  prepend<Before>(fn: (_: Before) => E): Event<Before>,
  subscribe(subscriber: Subscriber<E>): Subscription,
  /*::+*/ kind: kind,
  getType(): string,
  shortName: string,
  graphite: Graph,
  compositeName: CompositeName,
}
