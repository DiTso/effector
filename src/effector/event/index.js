//@flow

export {eventFabric} from './eventFabric'
export {forward} from './forward'
export {createEvent} from './createEvent'
export type {Event} from './index.h'
export {fromObservable} from './fromObservable'
export {filter} from './filter'
