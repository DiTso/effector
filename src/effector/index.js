//@flow

/*
 * People assume that time is a strict progression of cause to effect,
 * but actually from a non-linear, non-subjective viewpoint -
 * it's more like a big ball of wibbly wobbly time-y wimey stuff
 */

export {combine} from './combine'
export {sample} from './sample'

export type {Event} from 'effector/event'
export {createEvent, forward, fromObservable, filter} from 'effector/event'

export type {Effect} from 'effector/effect'
export {createEffect} from 'effector/effect'

export type {Store} from 'effector/store'
export {
  createStore,
  createStoreObject,
  setStoreName,
  createApi,
  restore,
  restoreEvent,
  restoreEffect,
  restoreObject,
} from 'effector/store'

export {Kind, clearNode, createGraph as createNode, step} from 'effector/stdlib'
export {launch} from 'effector/kernel'
export type {kind} from 'effector/stdlib'

export {
  invariant,
  warning,
  isUnit,
  isStore,
  isEvent,
  isEffect,
  isDomain,
  is,
} from 'effector/validate'
export {version} from 'effector/flags'

//eslint-disable-next-line
export * as blocks from 'effector/blocks'
