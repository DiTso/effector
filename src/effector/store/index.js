//@flow

export {storeFabric} from './storeFabric'
export {normalizeConfig} from '../config'
export {createStore} from './createStore'
export {createStoreObject} from './createStoreObject'
export {createApi} from './createApi'
export {restore, restoreEvent, restoreEffect, restoreObject} from './restore'

export type {Store} from './index.h'
