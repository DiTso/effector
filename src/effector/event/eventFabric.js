//@flow
import $$observable from 'symbol-observable'

import {
  step,
  Kind,
  stringRefcount,
  createGraph,
  type Unit,
} from 'effector/stdlib'
import type {Effect} from 'effector/effect'
import {launch} from 'effector/kernel'
import {noop, updateHandler} from 'effector/blocks'

import type {Subscription} from '../index.h'
import type {EventConfigPart} from '../config'
import type {Event} from './index.h'
import {createName} from '../compositeName'
import {forward} from './forward'

const nextID = stringRefcount()

export function eventFabric<Payload>({
  name: nameRaw,
  config = {},
}: {
  name?: string,
  config?: EventConfigPart,
}): Event<Payload> {
  const id = nextID()
  const name = nameRaw || id
  const compositeName = createName(name)
  const fullName = compositeName.fullName
  const graphite = createGraph({
    node: [
      step.emit({
        fullName,
      }),
    ],
  })

  //$off
  const instance: Event<Payload> = (payload: Payload) =>
    instance.create(payload)
  ;(instance: any).getType = () => fullName
  //eslint-disable-next-line no-unused-vars
  ;(instance: any).create = launch.bind(null, instance)
  ;(instance: any).kind = Kind.event
  ;(instance: any)[$$observable] = () => instance
  ;(instance: any).id = id
  ;(instance: any).watch = watchEvent.bind(null, instance)
  ;(instance: any).map = mapEvent.bind(null, instance)
  ;(instance: any).filter = filterEvent.bind(null, instance)
  ;(instance: any).prepend = prepend.bind(null, instance)
  ;(instance: any).subscribe = subscribe.bind(null, instance)
  instance.graphite = graphite
  instance.shortName = name
  instance.compositeName = compositeName
  instance.defaultConfig = config

  return instance
}

const subscribe = (event, observer): Subscription =>
  event.watch(payload => observer.next(payload))

const prepend = (event, fn: (_: any) => *) => {
  const contramapped: Event<any> = eventFabric({})
  forward({
    from: contramapped,
    to: createGraph({
      child: [event],
      scope: {handler: fn},
      node: [updateHandler],
    }),
  })
  return contramapped
}

declare function mapEvent<A, B>(event: Event<A>, fn: (_: A) => B): Event<B>
declare function mapEvent<A, B>(
  effect: Effect<A, any, any>,
  fn: (_: A) => B,
): Event<B>
function mapEvent<A, B>(event: Event<A> | Effect<A, any, any>, fn: A => B) {
  const mapped = eventFabric({})
  forward({
    from: event,
    to: createGraph({
      child: [mapped],
      scope: {handler: fn},
      node: [updateHandler],
    }),
  })
  return mapped
}

function filterEvent<A>(event: Unit, fn: A => boolean): Event<A> {
  const mapped = eventFabric({})
  forward({
    from: event,
    to: createGraph({
      scope: {handler: fn},
      child: [mapped],
      node: [
        step.filter({
          fn: (upd, {handler}) => !!handler(upd),
        }),
      ],
    }),
  })
  return mapped
}

function watchEvent<Payload>(
  event: Unit,
  watcher: (payload: Payload) => any,
): Subscription {
  return forward({
    from: event,
    to: createGraph({
      scope: {handler: watcher},
      //prettier-ignore
      node: [
        noop,
        step.run({
          fn: (payload: Payload, {handler}) => handler(
            payload,
          ),
        }),
      ]
    }),
  })
}
