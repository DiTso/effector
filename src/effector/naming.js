//@flow

import type {CompositeName} from './compositeName'
import {is} from 'effector/validate'
import type {Store} from 'effector/store'
import type {Event} from 'effector/event'
import type {Effect} from 'effector/effect'

export function getDisplayName(unit: {
  compositeName?: CompositeName,
  /*::+*/ id: string,
  /*::...*/
}) {
  if (unit.compositeName) {
    return unit.compositeName.fullName
  }
  return unit.id
}

const unitObjectMaxNames = 25

export function unitObjectName(
  objOrArr:
    | $ReadOnlyArray<Store<any> | Event<any> | Effect<any, any, any> | any>
    | {
        [key: string]: Store<any> | Event<any> | Effect<any, any, any> | any,
      },
) {
  let i = 0
  const arr: Array<any> = Object.values(objOrArr)
  const max = unitObjectMaxNames - 1
  const maxLength = arr.length - 1
  let name = 'combine('
  for (const unit of arr) {
    const comma = i === max || maxLength === i ? '' : ', '
    if (is.store(unit) || is.event(unit) || is.effect(unit)) {
      name += getDisplayName(unit) + comma
    } else {
      name += unit.toString() + comma
    }
    i += 1
    if (comma === '') break
  }
  name += ')'
  return name
}
