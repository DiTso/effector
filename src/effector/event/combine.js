//@flow

import type {Event} from './index.h'
import {eventFabric} from './eventFabric'

export function combine<T>(
  f: (...args: any[]) => T,
  ...events: $ReadOnlyArray<Event<any>>
): Event<T> {
  const unit = eventFabric({})
  const l = events.length
  let awaiting = l
  const values = new Array(l)
  const hasValue = new Array(l).fill(false)
  for (const [index, event] of events.entries()) {
    event.watch(value => {
      if (awaiting > 0) {
        if (!hasValue[index]) {
          hasValue[index] = true
          awaiting -= 1
        }
      }
      values[index] = value
      if (awaiting === 0) {
        unit(f(...values))
      }
    })
  }
  return unit
}
