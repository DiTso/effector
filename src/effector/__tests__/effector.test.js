//@flow

import {from, periodic} from 'most'

import {combine} from '..'

import {createDomain} from 'effector/domain'
import {createEvent, fromObservable} from 'effector/event'
import {createStore, createStoreObject} from 'effector/store'

test('will run in expected order', () => {
  const spy = jest.fn()
  const reset = createEvent('reset')
  const add = createEvent('add')
  const mult = createEvent('mult')
  const listSize = createStore(3)
    .on(add, (n, nn) => n + nn)
    .on(mult, (n, q) => n * q)
    .reset(reset)

  // const halt = add.link(
  //  mult, n => n%10, n => n + 10
  // )
  const currentList = createStore([])
    .on(add, (list, pl) => [...list, {add: pl}])
    .on(mult, (list, pl) => [...list, {mult: pl}])
    .reset(reset)
  const selected = createStore([])

  createStoreObject({listSize, currentList, selected})

  const unsub = currentList.subscribe(state => spy(state))
  add(5)
  mult(4)
  unsub()
  // halt()

  expect(spy.mock.calls).toEqual([[[]], [[{add: 5}]], [[{add: 5}, {mult: 4}]]])
  expect(spy).toHaveBeenCalledTimes(3)
})

test('reducer defaults', () => {
  const fn1 = jest.fn()
  const fn2 = jest.fn()
  const fn3 = jest.fn()
  const add = createEvent('add')
  const sub = createEvent('sub')
  const state1 = createStore(3)
    .on(add, (state, payload, type) => {
      fn1(state, payload, type)
    })
    .on(sub, (state, payload, type) => {
      fn2(state, payload, type)
      return state - payload
    })
  state1.watch(fn3)
  sub(1)
  add(10)
  add(2)
  expect({
    add: fn1.mock.calls,
    sub: fn2.mock.calls,
    watch: fn3.mock.calls,
    state: state1.getState(),
  }).toMatchSnapshot()
})

test('store.reset(event)', () => {
  const spy = jest.fn()
  const reset = createEvent('reset')
  const inc = createEvent('inc')
  const listSize = createStore(3)
    .on(inc, n => n + 1)
    .reset(reset)
  const currentList = createStore(
    Array.from({length: listSize.getState()}, (_, n) => n),
  )
    .on(inc, list => [...list, list.length])
    .reset(reset)
  const selected = createStore([])

  createStoreObject({listSize, currentList, selected})

  const unsub = currentList.subscribe(state => spy(state))
  inc()
  reset()
  unsub()

  expect(spy.mock.calls).toEqual([[[0, 1, 2]], [[0, 1, 2, 3]], [[0, 1, 2]]])
  expect(spy).toHaveBeenCalledTimes(3)
})

test('combine', () => {
  const spy = jest.fn()
  const inc = createEvent('inc')
  const dec = createEvent('dec')
  const s1 = createStore(0)
  const s2 = createStore(0)
  const s3 = createStore(0)
  const s4 = createStore(0)
  const result = combine(s1, s2, s3, s4, (a, b, c, d) => ({a, b, c, d}))
  result.watch(spy)
  s1.on(inc, _ => _ + 1).on(dec, _ => _ - 10)
  s2.on(inc, _ => _ + 10).on(dec, _ => _ - 1)

  expect(result.getState()).toMatchObject({a: 0, b: 0, c: 0, d: 0})

  inc()
  dec()
  expect(result.getState()).toMatchObject({a: -9, b: 9, c: 0, d: 0})

  expect(spy).toHaveBeenCalledTimes(3)
  // expect(fn).toHaveBeenCalledTimes(5)
})

test('no dull updates', () => {
  const store = createStore(false)
  const e1 = createEvent('e1')
  const e2 = createEvent('e2')
  const fn1 = jest.fn()
  const fn2 = jest.fn()
  const fn3 = jest.fn()
  store.watch(fn1)
  store.on(e1, (_, payload): boolean => payload)
  store.on(e2, (_, p) => _ === p)
  const nextStore = store.map(x => (fn2(x), x))
  nextStore.watch(fn3)
  store.watch(e => {})
  // nextStore.watch(e => console.log('next store', e))
  e1(false)
  e1(true)
  e1(false)
  e2(false)
  e2(false)
  expect(fn1.mock.calls).toMatchSnapshot()
  expect(fn2.mock.calls).toMatchSnapshot()
  expect(fn3.mock.calls).toMatchSnapshot()
  expect(fn1).toHaveBeenCalledTimes(5)
  expect(fn2).toHaveBeenCalledTimes(5)
  expect(fn3).toHaveBeenCalledTimes(5)
})

test('smoke', async() => {
  const used = jest.fn(x => Promise.resolve(x))
  const usedDone = jest.fn(x => Promise.resolve(x))
  const domain = createDomain('smoke')

  const effect = domain.effect('eff')
  effect.use(used)
  effect.done.watch(usedDone)
  const event = domain.event('event1')
  expect(effect).toBeDefined()
  expect(event).toBeDefined()
  event('bar')
  await effect('foo')
  expect(used).toHaveBeenCalledTimes(1)
  expect(usedDone).toHaveBeenCalledTimes(1)
})

//TODO Add port throws handling
describe('port', () => {
  test('port should work correctly', async() => {
    const used = jest.fn(state => {})
    const usedEff = jest.fn(state => {})
    const domain = createDomain()
    const event = domain.event('port-event')
    const eff = domain.event('port-effect')
    event.watch(used)
    eff.watch(usedEff)
    const str$ = periodic(100)
      .scan((a /*, b*/) => a + 1, 0)
      .take(10)

    // .map(event)

    str$.map(event).drain()
    await new Promise(rs => setTimeout(rs, 1500))
    expect(used).toHaveBeenCalledTimes(10)

    str$.map(eff).drain()
    await new Promise(rs => setTimeout(rs, 1500))
    expect(usedEff).toHaveBeenCalledTimes(10)
  })
})

it('works with most use cases', async() => {
  const spy = jest.fn()
  const timeout = createEvent('timeout')
  timeout.watch(spy)

  await periodic(300)
    .take(5)
    .observe(() => timeout())

  expect(spy).toHaveBeenCalledTimes(5)
})

test.skip('fromObservable supports own events as sources', async() => {
  const used = jest.fn(x => Promise.resolve(x))
  const usedDone = jest.fn(x => Promise.resolve(x))
  const domain = createDomain()

  const effect = domain.effect('eff')
  effect.use(used)
  effect.done.watch(usedDone)
  const event = domain.event('event1')
  fromObservable<string>(event).watch(e => effect(e))
  await event('ev')
  expect(used).toHaveBeenCalledTimes(1)
  expect(usedDone).toHaveBeenCalledTimes(1)
})
declare var epic: Function
//TODO WTF?
test.skip('hot reload support', async() => {
  // const fn = jest.fn()
  const fnA = jest.fn()
  const fnB = jest.fn()
  const used = jest.fn(x => Promise.resolve(x))
  const usedDone = jest.fn(x => Promise.resolve(x))
  const domain = createDomain()
  const storeA = domain.store({foo: 'bar'})
  storeA.watch((s, x) => (fnA(x), s))

  const effect = domain.effect('eff')
  effect.use(used)
  effect.done.watch(usedDone)
  const event = domain.event('event1')
  epic(event, data$ => data$.map(e => effect(e)))
  await event('ev')
  expect(used).toHaveBeenCalledTimes(1)
  expect(usedDone).toHaveBeenCalledTimes(1)

  const storeB = domain.store({foo: 'bar'})
  storeB.watch((s, x) => (fnB(x), s))

  await event('ev')
  expect(used).toHaveBeenCalledTimes(2)
  expect(usedDone).toHaveBeenCalledTimes(2)
  expect(fnA).toHaveBeenCalledTimes(5)
  expect(fnB).toHaveBeenCalledTimes(4)
})

/*
test('typeConstant', async() => {
 const fn = jest.fn()
 const used = jest.fn((x: string) => console.log(x))
 const respFn = jest.fn(x => console.log(x))
 const domain = createDomain('with-prefix')
 const store = domain.store({foo: 'bar'})
 const event = domain.typeConstant('TYPE_CONST')
 const eventResp = domain.typeConstant('RESP')
 eventResp.watch(respFn)
 expect(event.getType()).toBe('TYPE_CONST')
 event.epic(data$ =>
  data$.map(e => {
   used(e)
   return eventResp(e)
  }),
 )

 log`type constant's event`(event)
 store.dispatch({type: 'TYPE_CONST', payload: 'raw'})
 expect(event('test')).toMatchObject({type: 'TYPE_CONST', payload: 'test'})
 expect(event).toBeDefined()
 await event('run')
 await delay(500)
 expect(respFn).toHaveBeenCalledTimes(2)
 expect(used).toHaveBeenCalledTimes(2)
 expect(used.mock.calls).toMatchObject(
  //$ off
  [['raw'], ['run']],
 )
})
*/
test('subscription', async() => {
  const spy = jest.fn()
  const domain = createDomain()

  const eff = domain.effect('TYPE_CONST')
  expect(() => {
    from(eff).observe(spy)
  }).not.toThrow()
  const event = domain.event('ev')
  expect(() => {
    from(event).observe(spy)
  }).not.toThrow()
  await event('')
  await eff('')
  expect(spy).toHaveBeenCalledTimes(2)
})
