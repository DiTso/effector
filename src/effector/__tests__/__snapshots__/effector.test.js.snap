// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`no dull updates 1`] = `
Array [
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
]
`;

exports[`no dull updates 2`] = `
Array [
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
]
`;

exports[`no dull updates 3`] = `
Array [
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
]
`;

exports[`reducer defaults 1`] = `
Object {
  "add": Array [
    Array [
      2,
      10,
      undefined,
    ],
    Array [
      2,
      2,
      undefined,
    ],
  ],
  "state": 2,
  "sub": Array [
    Array [
      3,
      1,
      undefined,
    ],
  ],
  "watch": Array [
    Array [
      3,
    ],
    Array [
      2,
    ],
  ],
}
`;
